Style create/join pages and improve game page UI

alankbi opened the pull request:
### List of Changes:

* Apply circle background and input styling to join and create pages
* Refactor circle background to use absolute positioning based on entire screen
  * This was to fix the issue on create/join pages where the popup text at the bottom changes the content height and causes the circles to jump down and up
* Update language select design to be compatible with Safari/Firefox
  * Unfortunately, there's no easy way to keep the arrow if we want it to look nice on those other browsers
* Apply custom styling to markdown and code editors to improve the UI (and add a small footer)
* Edit console to display run/submit output in a nice format

### Notes:

Couple things I didn't add (actually a lot of things):

* Difficulty tag as seen in the Figma
  * Since the colored tags are implemented in #115, I figured whichever of the two PRs merges last could just push an extra commit to add the tag below the problem title
* Drag icons between panels 
  * BinarySearch does this by styling a div inside the separator panel. However, idk how we'd be able to do this without doing something sketchy like using JS to like manually inject a custom-styled div inside the splitter library's separator class
* Timer absolutely positioned at top right of problem pane (as seen if Figma)
  * Personally I think it looks good where it is now (header bar), and moving it to the problem pane might cause the same absolutely positioned overlap errors (when the problem name is long) that we currently get with the run/submit code buttons
* Gradient outlines for the textareas
  * I don't think this is actually 100% feasible with textareas or input tags - the two things I got to work either didn't support border radii or didn't support padding (see below):

![No padding inside textarea with gradient outline](https://user-images.githubusercontent.com/24768574/111260080-6beea280-85dd-11eb-85ec-9af71d925269.png)

Ultimately though if there are any things among that list that we think we should add for the MVP let me know and I can add that to this PR.

### Screencast and Images:

[Screencast (2x speed recommended)](https://www.loom.com/share/e27554e3130b4969aa5d67d39a6e489c)

alankbi commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Update console display and game page finetuning](https://trello.com/c/tnHnMAyb/297-update-console-display-and-game-page-finetuning)

zpChris commented on frontend/src/components/game/Editor.tsx:
```diff
@@ -63,6 +80,18 @@ function ResizableMonacoEditor(props: EditorProps) {
+      inherit: true,
+      rules: [],
+      colors: {
+        'editor.lineHighlightBorder': '#f8f8f8',
+        'editor.lineHighlightBackground': '#f8f8f8',
```
nit: Add or change these colors to ones in the `Theme.tsx` file.

zpChris commented on frontend/src/components/game/Editor.tsx:
```diff
@@ -98,13 +127,29 @@ function ResizableMonacoEditor(props: EditorProps) {
       <EditorContainer>
         <MonacoEditor
           options={{
+            automaticLayout: true,
             fixedOverflowWidgets: true,
+            fontFamily: 'Monaco',
+            hideCursorInOverviewRuler: true,
             minimap: { enabled: false },
-            automaticLayout: true,
+            overviewRulerBorder: false,
+            overviewRulerLanes: 0,
+            padding: { top: 10, bottom: 10 },
+            quickSuggestions: false,
+            renderLineHighlight: 'all',
+            renderIndentGuides: false,
+            renderWhitespace: 'none',
+            scrollbar: {
+              verticalScrollbarSize: 5,
+              horizontalScrollbarSize: 5,
+              useShadows: false,
+            },
             scrollBeyondLastLine: false,
           }}
```
nit: Define the options in a separate variable and put that variable here.

alankbi replied:
Done - will note that I had to add a type any to the separate variable in order to get it to be compatible with this options parameter 

I think another comment mentioned the use of any so I can look into if types exist when going over that one

zpChris replied:
No problem at all, yeah I think the specific types only apply when it's reasonably easy, for some of the external libraries it's tough to find the types.

zpChris commented on frontend/src/components/game/Editor.tsx:
```diff
@@ -63,6 +80,18 @@ function ResizableMonacoEditor(props: EditorProps) {
     });
   };
 
+  const handleEditorWillMount = (editor: any) => {
```
nit: Is the editor any specific type? If so, add that type here - no problem if this isn't straightforward to find

alankbi replied:
Did a little digging and was able to replace all the anys in this class - the types are pretty weird tho, like `typeof monaco` or `monaco.editor.IStandaloneCodeEditor`, so lmk what your thoughts are about that

zpChris replied:
I think that's good - thanks for doing that digging!

zpChris commented on frontend/src/components/layout/CircleBackground.tsx:
```diff
@@ -3,11 +3,13 @@ import styled from 'styled-components';
+  position: relative;
   width: 100%;
-  min-height: 100vh;
+  height: 100vh;
```
I remember that I ran into issues on my PR with using vw - is there any need to change this or no? Not a requested change lol, just genuinely asking

alankbi replied:
The idea behind this was to have the circle positioning depend on the entire view size of the screen - however ik you brought up the circles in the screencast so we can also talk about it there (whether it should be changed to static to keep constant regardless of screen size for example)

zpChris commented on frontend/src/components/problem/ProblemDisplay.tsx:
```diff
@@ -271,8 +271,8 @@ function ProblemDisplay(props: ProblemDisplayParams) {
 
-      <LargeInputButton value={actionText} onClick={() => onClick(newProblem)} />
-      {editMode ? <LargeInputButton value="Delete Problem" onClick={deleteProblemFunc} /> : null}
+      <PrimaryInput value={actionText} onClick={() => onClick(newProblem)} />
+      {editMode ? <PrimaryInput value="Delete Problem" onClick={deleteProblemFunc} /> : null}
```
This is a good change, thanks! Just wanted to note that you'll run into merge conflicts I think, so this may have to be updated

alankbi replied:
Just fixed - I think I could just delete this refactor since the buttons not being used there anymore anyways

zpChris commented on frontend/src/components/layout/CircleBackground.tsx:
```diff
@@ -31,49 +33,51 @@ function CircleBackgroundLayout({ children }: MyProps) {
+        <CircleBackgroundContainer>
+          {children}
+        </CircleBackgroundContainer>
+        <div>
```
I made a screencast with some of my personal design nits / preferences: [Design Nits on #118 Frontpages](https://www.loom.com/share/3e8affa5e05f43cdb67c4db248bc4acb). I think that the circle background isn't entirely aligned on the other frontpages.

alankbi replied:
I made a slight fix to this by slightly lowering the top of the containers for the create and join pages. I don't think we'll be able to get it perfect on every screen size though, for example: 

![Small width screen with overlapping text](https://user-images.githubusercontent.com/24768574/111862199-ffdba980-8910-11eb-85a5-0630615da922.png)

Any thoughts?

zpChris replied:
It's not perfect, but I don't have a huge problem with these designs, especially if you like this more than just the plain background on these pages. It also means less development on the likely slightly less important tasks here, which should be a good thing.

Maybe in the future, we could have this layout depend again on the central container rather than the whole page, and then each of these pages would be customized to that. If the circles / blobs moved fluidly (animation-wise) when transitioning the pages, that would be the optimal design in my opinion; however, this may be tough (or even infeasible), and I'm ok with this setup. We could look at it again in the future, but if this works by you for this PR, it's all good by me.

Also, thanks for making the buttons the same size as the input ❤️

zpChris commented on frontend/src/components/core/EnterNickname.tsx:
```diff
@@ -79,7 +79,7 @@ export default function EnterNicknamePage(props: EnterNicknameProps) {
         }}
       />
-      <LargeInputButton
+      <PrimaryInput
```
I also talked about this in [this screencast](https://www.loom.com/share/3e8affa5e05f43cdb67c4db248bc4acb) (referenced elsewhere as well) - I think this page looks best when the button matches the input size, and when the button is able to be "disabled" as well, but of course open to a discussion on this.

alankbi replied:
Updated

zpChris commented on frontend/src/views/Game.tsx:
```diff
@@ -34,10 +34,16 @@ import {
+import { NoColorExternalLink } from '../components/core/Link';
 import Language from '../api/Language';
 
 const StyledMarkdownEditor = styled(MarkdownEditor)`
```
So with reference to [this discussion](https://github.com/rocketden/main/pull/114#issuecomment-799796479), I saw that the game page has lines at those locations even when there isn't an overflow of users - is there a way to remove those lines if the overflow doesn't exist?

<img width="1050" alt="Shadow Remaining on Edges of Player List" src="https://user-images.githubusercontent.com/7987279/111438238-3de28e80-86c1-11eb-81f8-9c4f950e22d6.png">

alankbi replied:
I played around with this a bit and it seems like it's a thin line that shows up every now and then on very specific browser widths. I couldn't come up with a perfect solution other than setting the background shadow position to like 0.1% and 99.9% so that the cover can completely hide it even with off by one browser pixels - not the best solution lol and it might not handle 100% of the unique browser widths but hopefully should capture the majority of them - lmk what you think

zpChris replied:
Huh that's a really weird bug - yeah I'm ok with this solution, we can revisit this later if need be to come up with a more rigorous solution.

zpChris reviewed (changes requested):
Requesting changes because there are some design changes / ideas I have in mind, as well as various nits / requested changes on the PR.

I didn't talk too much about the updated game page as I think the code style itself is great, but I made a screencast with potential design changes / additions here: [Redesigned Game Page Feedback](https://www.loom.com/share/6fd93239b960498395676f74981d28bf).

zpChris commented on frontend/src/views/Game.tsx:
```diff
@@ -409,6 +415,10 @@ function GamePage() {
             />
+            <BottomFooterText>
+              {'Notice an issue? Contact us at '}
+              <NoColorExternalLink href="mailto:support@codejoust.co">support@codejoust.co</NoColorExternalLink>
```
Check out #119! I've added "copy email" functionality there, I think that could potentially be incorporated into this PR (we can likely merge #119 first since that's a simpler PR, but let me know your thoughts).

alankbi replied:
Sounds good - I'll wait for that PR to merge before implementing it then (since I just approved it as well)

zpChris replied:
I just merged that PR, sounds good 👍 

The "Email copied!" text may overlap with the time in this case, not sure if anything should be done about that (likely not, idk), but just wanted to note that.

alankbi replied:
Done! I moved the copy popup to the bottom to avoid the overlap.

zpChris replied:
Ayyy nice, looks great! I'll note that it has an animation up but then disappears immediately when you close it, but that doesn't actually bother me as I think that looks fine too. (The only note may be a lack of consistency with other pages, but it's already on the bottom of the page instead of the top, so it's not exactly the same anyway.)

alankbi commented on frontend/src/views/Game.tsx:
```diff
@@ -34,10 +34,16 @@ import {
 const StyledMarkdownEditor = styled(MarkdownEditor)`
   padding: 0;
+  
+  p {
+    color: ${({ theme }) => theme.colors.text};
+    font-family: ${({ theme }) => theme.font};
+  }
```
Thanks for the detailed writeup haha - sounds good to me

alankbi commented on frontend/src/components/game/Editor.tsx:
```diff
@@ -15,7 +15,8 @@ const Content = styled.div`
   flex-direction: column;
-  padding: 0.5rem 0;
+  padding: 0;
+  margin: 0 -5px;
```
Sure - I got rid of this and instead reduced the padding of the surrounding panel to 0. Note that now there's no padding in between so this is what is looks like: 

![No padding on editor in panel](https://user-images.githubusercontent.com/24768574/111739702-b885e800-8840-11eb-8a73-28a671d6c3e7.png)

I'm also open to adding back a little extra padding too depending on what looks best

zpChris replied:
I'm assuming the padding effects show on the scrollbar - I actually like that more as now it doesn't overlap with the language button, so no problem at all 👍

alankbi commented:
Made the changes described above and in the screencasts. For the game page screencast:

Time scroll:

I updated this to use the same scrolling feature as for the nickname. The UI here isn't perfect for sure tho since it can sometimes cut off at weird points, but lmk how you feel about that/any potential solutions if it's bothering at all

Center language select text:

I implemented this to work on both Chrome and Firefox. However, the top solutions [here](https://stackoverflow.com/questions/10813528/is-it-possible-to-center-text-in-select-box) don't seem to work for Safari. 

Console resize: 

I re-added the ability to resize in both directions - this means that the auto-enlarge/retracting feature of the text areas will work up until the user decides to manually change the size (and then from there the size will be constant unless they resize it again). 

As for the resize bug with the handle disappearing, I actually visually indicated this in the screencast haha but forgot to bring it up - I tried finding a solution to it before and tried again a bit but nothing seemed to work and idk why - of note would be [this](https://bugs.chromium.org/p/chromium/issues/detail?id=94583) thread on the same bug. Any ideas/solutions? I'm stuck lol

Save input: 

Yeah, I think that makes sense. I'll add a card to the backlog for this and the whole multiple panel thing in the console which might be a good addition maybe in the future though

alankbi commented:
Re-requesting review although there were lots of comments so feel free to lmk if I missed anything :)

alankbi commented:
Also just realized that there's a +17k line change somewhere in here - I'll go ahead and figure out where + create a new branch lol

zpChris commented:
> Re-requesting review although there were lots of comments so feel free to lmk if I missed anything :)
> 
> Also as talked about, I pushed code in this PR to change our Heroku repo (and the deploy steps, though untested), and also added Google Analytics (you should have received an email for collaborator access to that account).

Thanks so much for making those additions! Much appreciated 😊

zpChris commented:
> Also just realized that there's a +17k line change somewhere in here - I'll go ahead and figure out where + create a new branch lol

If you can't figure out what's causing that line increase, I'm ok just reviewing this PR, no worries either way 👍

zpChris commented on frontend/src/components/core/Input.tsx:
```diff
@@ -1,9 +1,14 @@
+  height?: string,
+};
+
+const Input = styled.input<Dimensions>`
```
I think all implementations of `Input` should implement the `font-family: ${({ theme }) => theme.font}` (the last one is currently not, but I think it should be added!). Could you add that line to this styled component, and remove it from the below ones?

alankbi replied:
Sounds good, added - I left it in though for PureTextInputTitle since that one doesn't inherit Input (and changing it to inherit might cause some of the styles to clash)

zpChris commented on frontend/src/components/core/Input.tsx:
```diff
@@ -30,49 +35,77 @@ export const PureTextInputTitle = styled.input.attrs(() => ({
+  white-space: pre;
+  resize: both;
+  
+  /* Hide scrollbar for Chrome, Safari and Opera */
```
How do you test all of this - is there a way to check cross-browser functionality quickly?

alankbi replied:
This comment is from the post that gave the solution code - as for testing, I don't have a quick method besides just doing the manual testing that we usually do on Chrome/Safari/Firefox. For clarity though I'll remove the Opera from this comment

zpChris commented on frontend/src/views/ContactUs.tsx:
```diff
@@ -20,7 +20,7 @@ function ContactUsPage() {
         </CopyIndicator>
       </CopyIndicatorContainer>
-      <LandingPageContainer>
+      <CircleBackgroundContainer>
```
So this page doesn't have a circle background, so this name is a bit confusing (I thought you added that background to this page, though I think it looks better without for this page). Maybe rename that container or something?

alankbi replied:
Haha yeah that can be a bit confusing - I renamed to DynamicWidthContainer - lmk if you have any alternative ideas for a name though

zpChris reviewed (changes requested):
I left a lot of comments / questions / nits, but nothing huge, so this should be quick to implement. I wanted to note the future changes that we'll make to the Game page, for planning purposes (only for future PRs, not this):
- Move the "Loading..." / error message that occurs on run or submit so that the game page doesn't all shift down.
- Perhaps update the styling of the "notifications", though I'm open to discussion on this.
- Save the previous test case before "Run Code" somehow, perhaps through different tabs on the bottom-right panel.
- Add the "Difficulty" button beside the name (note that the title can be multiple lines if it is long / panel is collapsed, so some styling decisions should be made here).
- Perhaps the "Run Code" and "Submit" buttons are fixed to the bottom of the bottom-left panel, so the input / output / console input boxes are all full-width.
- A modal pop-up that appears when time is up, where they can then click to go to the Results page. They could perhaps also keep coding on this problem if they wanted, or use that game page view to see other people's code.
- Similar to binarysearch, we could add "tabs" to the code panel to view other people's blurred-out code (and with powerups, their actual code). We'll have to figure out how live updates would work here.
- Possibly more, like a chat, powerups, teams, and more - but that's all for the future!

alankbi commented on .github/workflows/workflow.yml:
```diff
@@ -31,5 +31,5 @@ jobs:
-          heroku_app_name: "rocketden"
-          heroku_email: "zpchris@wharton.upenn.edu"
+          heroku_app_name: "codejoust"
+          heroku_email: "support@codejoust.co"
```
Yeah I wouldn't think so, but this is also untested (since we haven't merged it into main yet) so it's possible an error could come up. But yeah leaning towards no

alankbi commented on frontend/src/components/config/App.tsx:
```diff
@@ -17,13 +18,23 @@ import CircleBackgroundLayout from '../layout/CircleBackground';
 import ProblemLayout from '../layout/ProblemLayout';
 
+// Set up Google Analytics
+ReactGA.initialize('UA-192641172-2');
```
I think it should be okay/expected to be public - IIRC you'd be able to see it in the browser console (since it's like an embedded script) so it wouldn't be secure anyways.

alankbi commented on frontend/src/components/core/Button.tsx:
```diff
@@ -21,12 +26,12 @@ export const DefaultButton = styled.button`
   color: ${({ theme }) => theme.colors.white};
-  width: ${({ width }) => width || '10rem'};
-  height: ${({ height }) => height || '2.75rem'};
+  width: ${({ width }) => width || '12rem'};
+  height: ${({ height }) => height || '3rem'};
```
Haha nice catch (for both of them) - I think the slightly bigger button is fine on the problem display but I'm open to changing if you prefer the original (or on your PR too)

zpChris replied:
That change is fine, no problem 👍

alankbi commented on frontend/src/components/core/Input.tsx:
```diff
@@ -1,9 +1,14 @@
   box-sizing: border-box;
   border-radius: 0.25rem;
-  border: 3px solid ${({ theme }) => theme.colors.blue};
+  border: 2px solid ${({ theme }) => theme.colors.blue};
```
Hm yeah that might be useful at some point (or even just adding the entire border definition into theme - along the same lines we could also add the box-shadow to theme as well since ik that one is pretty heavily copied into various places

alankbi commented on frontend/src/components/game/Editor.tsx:
```diff
@@ -30,30 +32,73 @@ const LanguageContainer = styled.div`
-  font-style: italic;
+  color: ${({ theme }) => theme.colors.text};
+  text-align: center;
+  text-align-last: center;
```
This centers the text in the language select for Chrome: https://stackoverflow.com/a/34532555/7386515

alankbi commented on frontend/src/views/Game.tsx:
```diff
@@ -46,6 +62,10 @@ const OverflowPanel = styled(Panel)`
   padding: 0 25px;
 `;
 
+const LowPaddingPanel = styled(Panel)`
```
leaving unresolved until i remember to refactor the usage of this to the new name

alankbi commented on frontend/src/views/Game.tsx:
```diff
@@ -63,7 +83,7 @@ const LeaderboardContent = styled.div`
     ${({ theme: { colors: { background } } }) => `linear-gradient(to left, rgba(0,0,0,.12), ${background})`};
 
-  background-position: left center, right center, left center, right center;
+  background-position: left center, right center, 0.15% center, 99.85% center;
   background-repeat: no-repeat;
   background-color: ${({ theme }) => theme.colors.background};
   background-size: 20px 100%, 20px 100%, 10px 100%, 10px 100%;
```
image: the shadow linear gradients are the shadows you see on the leaderboard when there's scroll, whereas the shadow cover linear gradients are white (ish) shadows that cover those shadows when there's not scroll (there's 4 because of left and right + shadow and shadow cover
position: sets each of those four to be positioned at these places (left center would be 0% 50% i.e. left 0% and top 50%)
repeat: no idea, my guess is just makes it appear once rather than repeating itself
color: background color
size: for each of the four, set the width and height (i.e. 20px wide and 100% height within the leaderboard container)

this is just my understanding of it though, since the stack post I saw gave the solution but like no explanation behind it

alankbi commented on frontend/src/components/special/CopyIndicator.tsx:
```diff
@@ -33,8 +45,16 @@ export const InlineCopyText = styled(ContactHeaderText)`
   border-bottom: 1px solid ${({ theme }) => theme.colors.text};
 `;
 
+export const SmallInlineCopyText = styled(InlineCopyText).attrs({ as: 'p' })`
```
This makes it a p tag as opposed to an h1 tag (since InlineCopyText is an h1 tag I believe) which would fit the use case better on the game page

alankbi commented on frontend/src/components/special/CopyIndicator.tsx:
```diff
@@ -17,6 +17,18 @@ export const CopyIndicatorContainer = styled.div.attrs((props: CopyIndicator) =>
 
+export const BottomCopyIndicatorContainer = styled.div.attrs((props: CopyIndicator) => ({
+  style: {
+    transform: (!props.copied) ? 'translateY(60px)' : null,
+    visibility: (props.copied) ? 'visible' : 'hidden',
```
More like a slight side effect but yeah - I changed the button to be position fixed (in order to get it to work from below as it didn't look good at the top due to overlap) and added the visibility hidden so it wouldn't show up when not clicked

alankbi commented on frontend/src/components/core/Button.tsx:
```diff
@@ -2,7 +2,12 @@ import styled from 'styled-components';
+  height?: string,
+};
+
+export const DefaultButton = styled.button<Dimensions>`
```
Done!

alankbi commented:
Made the changes/responded to comments and re-requested review - and that list of future changes looks great!

zpChris reviewed (approved):
LGTM! Thanks for making all those changes 👍

alankbi commented:
This is the new look with the difficulty showing under the problem title - feel free to leave any suggestions for design here (if not in time before merging this PR and you have a major gripe, I can address it in the next results page PR)

I refrained from putting it next to the title as that would likely lead to overflow issues when the title is too long

![Easy difficulty tag](https://user-images.githubusercontent.com/24768574/112426978-1c535980-8cf6-11eb-9dc0-6edb4c587c1e.png)

zpChris commented:
> This is the new look with the difficulty showing under the problem title - feel free to leave any suggestions for design here (if not in time before merging this PR and you have a major gripe, I can address it in the next results page PR)
> 
> I refrained from putting it next to the title as that would likely lead to overflow issues when the title is too long
> 
> ![Easy difficulty tag](https://user-images.githubusercontent.com/24768574/112426978-1c535980-8cf6-11eb-9dc0-6edb4c587c1e.png)

Checking it out right now! I'll edit this comment.

alankbi commented:
> > This is the new look with the difficulty showing under the problem title - feel free to leave any suggestions for design here (if not in time before merging this PR and you have a major gripe, I can address it in the next results page PR)
> > I refrained from putting it next to the title as that would likely lead to overflow issues when the title is too long
> > ![Easy difficulty tag](https://user-images.githubusercontent.com/24768574/112426978-1c535980-8cf6-11eb-9dc0-6edb4c587c1e.png)
> 
> I think the button could be a tad bigger, and maybe since it's not displayed in a row with the other buttons, it could instead simply fit the size of the text rather than have a fixed width? @alankbi let me know what you think.

I can make it slightly higher taller for sure - can you clarify what you mean by making it fit the size of the text?

zpChris commented:
> I can make it slightly higher taller for sure - can you clarify what you mean by making it fit the size of the text?

Yeah so right now, no matter what text you put in the button, I believe it has a set width (correct me if I'm wrong). We did this because normally difficulty buttons are all displayed side-by-side; however, now that just one difficulty button will be shown here, you could potentially remove the `width` attribute and just add padding instead, so it better fits the text (whether that is "Easy", "Medium", or "Hard"). Thoughts?

alankbi commented:
Done!

![Smaller width difficulty button text](https://user-images.githubusercontent.com/24768574/112429915-a0a7db80-8cfa-11eb-8ed4-a3dc14721d9f.png)

![Larger width difficulty button text](https://user-images.githubusercontent.com/24768574/112429929-a7cee980-8cfa-11eb-8647-9d78ab189763.png)

zpChris commented:
> Done!
> 
> ![Smaller width difficulty button text](https://user-images.githubusercontent.com/24768574/112429915-a0a7db80-8cfa-11eb-8ed4-a3dc14721d9f.png)
> 
> ![Larger width difficulty button text](https://user-images.githubusercontent.com/24768574/112429929-a7cee980-8cfa-11eb-8647-9d78ab189763.png)

Yay love it! In the above image the "Hard" button is actually the "Medium" color haha, but I'm guessing you just edited the frontend text to show that it changes width based on the text.
